clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -3335,6 +3362,32 @@ SYCLIntelNumSimdWorkItemsAttr *Sema::MergeSYCLIntelNumSimdWorkItemsAttr(
       return nullptr;
     }
   }
+  // If the declaration has a [[intel::reqd_work_group_size()]]
+  // attribute, check to see if the last argument can be evenly
+  // divided by the num_simd_work_items attribute.
+  // If the declaration has a __attribute__((reqd_work_group_size))
+  // or [[cl::reqd_work_group_size()]] attribute, check to see if the
+  // first argument can be evenly divided by the num_simd_work_items
+  // attribute. GNU and [[cl::reqd_work_group_size()]] spelling of
+  // ReqdWorkGroupSizeAttr maps to the OpenCL semantics. First and last
+  // argument are only swapped for the Intel atributes in SYCL and not
+  // the OpenCL ones.
+  if (const auto *DeclAttr = D->getAttr<ReqdWorkGroupSizeAttr>()) {
+    const auto *MergeExpr = dyn_cast<ConstantExpr>(A.getValue());
+    const auto *DeclXExpr = dyn_cast<ConstantExpr>(DeclAttr->getXDim());
+    const auto *DeclZExpr = dyn_cast<ConstantExpr>(DeclAttr->getZDim());
+
+    llvm::APSInt WorkGroupSize = DeclAttr->usesOpenCLArgOrdering()
```

Comment:
Similar here -- I think this could crash because the work group size or the merge expression hasn't been evaluated yet (so we'd dereference a null pointer). I think on something like this:
```
template <int X, int Y, int Z, int N>
[[intel::reqd_work_group_size(X, Y, Z)]] void func();

template <int X, int Y, int Z, int N>
[[intel::num_simd_work_items(N)]] void func();
```

Reply:
I tried this example:
```
template <int X, int Y, int Z, int N>
[[intel::reqd_work_group_size(X, Y, Z)]] void func();

template <int X, int Y, int Z, int N>
[[intel::num_simd_work_items(N)]] void func();

int main() {
  func<3, 3, 5, 3>();
}

This does not crash.

test.cpp:5:3: error: 'num_simd_work_items' attribute must evenly divide the work-group size for the 'reqd_work_group_size' attribute
[[intel::num_simd_work_items(N)]] void func();
  ^
test.cpp:8:3: note: in instantiation of function template specialization 'func<3, 3, 5, 3>' requested here
  func<3, 3, 5, 3>();
  ^
test.cpp:2:3: note: conflicting attribute is here
[[intel::reqd_work_group_size(X, Y, Z)]] void func();
  ^
1 error generated.
```
Am i missing something here?

Reply:
This would crash. I think you meant to test this example. The crash happens here because the work group size or the merge expression hasn't been evaluated yet for existing implementation of reqd_work_group_size attribute( plan to refactor the current implementation)

template <int X, int Y, int Z>
[[intel::num_simd_work_items(3)]] void func();

template <int X, int Y, int Z>
[[intel::reqd_work_group_size(X, Y, Z)]]
void func();

int main() {
  func<3, 3, 5>();
}

Reply:
It's possible my test cases weren't quite right. The cases I'm worried about are when either the arg expression from `num_simd_work_items` is null (because it's not a constant expression we've evaluated yet) or the work group size argument from `reqd_work_group_size` is null (for the same reason).

(Note how we're checking for a null `MergeExpr` a few lines higher in this same function.)